ls1intum/Artemis
# `Development`: Improve client and server tests related to grading, bonus and exam

### Checklist
#### General


- [x] This is a small issue that I tested locally and was confirmed by ~another developer on a test server~ automated tests.
#### Server
#### Client

### Motivation and Context


There are several issues with the current tests noticed in the last weeks and we want to improve the quality of the client and server tests.

### Description

Client:
- Added tests for components below which were not tested:
  - grading-system.component.ts
  - mode-picker.component.ts
- Added tests to increase coverage:
   - exercise.service.ts
   - student-exam-detail.component.ts



Server:
- Added tests to increase coverage.
  - CourseScoreCalculationService.java
  - PlagiarismCase.java
  - BonusStrategy.java
  - BonusService.java
- Addressed TODOs below in ExamIntegrationTest.java:
  - Added assert statements to 3 test cases that are missing them.
  - Avoided duplicated code with StudentExamIntegrationTest  by extracting common parts to new methods and reusing.
  - Supported programming exercises in testGetExamScore.
  - Added a test case for which hasSecondCorrectionAndStarted is true.

### Steps for Testing

No Prerequisites.

1. Check client and server test results on Bamboo and GitHub Actions.
2. Check if tests contain reasonable assert/expect statements.

### Review Progress






#### Code Review

### Test Coverage



| Class/File | Branch | Line |
|------------|-------:|-----:|
| CourseScoreCalculationService.java | 77% | 94% |
| PlagiarismCase.java | 100% | 100% |
| BonusStrategy.java | 100% | 97% |
| BonusService.java | 90% | 100% |
| ExamService.java | 82% | 95% |
| grading-system.component.ts | 80% | 100% |
| mode-picker.component.ts | 100% | 100% |
| exercise.service.ts | 83.18% | 91.09% |
| student-exam-detail.component.ts | 65.68% | 93.1% |
